### PR TITLE
perf(concurrency): fast-path uncontended permits

### PR DESCRIPTION
**Pull request by thomhurst**
## Summary

- use atomic permit acquisition for the uncontended concurrency-limit path
- retain `SemaphoreSlim` only as the handoff queue for registered waiters
- preserve capacity, cancellation, metrics, and publication-failure accounting

## Profiling finding

An EventPipe CPU sample of the five-strategy chain attributed roughly one-third of sampled workload time to `Monitor.Enter_Slowpath` under `SemaphoreSlim.WaitCore` and `SemaphoreSlim.Release`. Even when a permit was immediately available, each execution entered the semaphore's monitor-backed path.

## Benchmark

BenchmarkDotNet 0.15.8, default job, .NET 10.0.11, Windows 11, Intel Core Ultra 9 185H. Both runs used the same checkout base and machine.

| Benchmark | Before | After | Change | Allocated |
|---|---:|---:|---:|---:|
| `Kevlar_Uncontended` | 312.3 ns | 218.3 ns | **30.1% faster (1.43x)** | 0 B -> 0 B |

Command:

```powershell
dotnet run --project benchmarks/Kevlar.Benchmarks -c Release --no-build -- --filter "*ConcurrencyLimitBenchmarks.Kevlar_Uncontended"
```

## Validation

- `dotnet build Kevlar.slnx -c Release`
- `dotnet run --project tests/Kevlar.Tests -c Release --no-build -- --timeout 5m` (612 passed, including concurrency race/edge cases)
- `dotnet run --project tests/Kevlar.AllocationTests -c Release --no-build -- --timeout 5m` (2 passed)

## Summary by CodeRabbit

* **Performance**
  * Improved concurrency-limit handling for uncontended operations, reducing unnecessary waiting overhead.
  * Uncontended asynchronous executions can now complete immediately while returning the expected results.

* **Reliability**
  * Improved coordination between synchronous and asynchronous waits.
  * Enhanced permit availability and handoff when concurrent operations complete.
  * Improved cancellation, permit release, and cleanup when operations fail to start.
  * Ensured permits remain immediately available for subsequent operations after release.

### CODE REVIEW

File: src/Kevlar/Strategies/ConcurrencyLimit/ConcurrencyLimitStrategy.cs [hunk 1]
```diff
@@ -163,12 +174,42 @@ public override async ValueTask<Outcome<T>> ExecuteAsync<T, TState>(Continuation
...
+                return true;
+            }
+        }
+    }
+
+    private void ReleasePermit()
+    {
+        // A waiter registers before retrying the atomic pool, so a concurrent release either
+        // replenishes that pool for the retry or signals the semaphore after the retry fails.
+        if (Volatile.Read(ref _waiters) > 0)
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply src/Kevlar/Strategies/ConcurrencyLimit/ConcurrencyLimitStrategy.cs]: Fixed in 59817dc. Release now publishes the atomic permit before checking registered waiters, then transfers it to the semaphore only if it can atomically reclaim it. A concurrent waiter therefore either wins the retry or receives a signal; the missed-wakeup window is closed. Build, 612 unit tests, allocation gates, and ShortRun (112.2 ns, 0 B) pass.
